### PR TITLE
Fix watch reconnect

### PR DESCRIPTION
**Pull request by mhmxs**
Post ready has to be less strict, to ensure node watcher starts as soon as possible.

Previously watcher tried to connect forever, but in some cases this fails forever. With this change Operator gives up watching node and operator config after a few tries and forces restart of operator.

### CODE REVIEW

File: watchers/common.go [hunk 1]
```diff
@@ -16,32 +15,41 @@ const maxWaitPeriod = 30
...
-func start(ctx context.Context, name string, watchFunc watchFunc) {
+func start(ctx context.Context, name string, watchFunc watchFunc) <-chan error {
+	errChan := make(chan error)
+
 	go func() {
 		log := ctrl.Log.WithName(name + " watcher")
 
-		for waitPeriod := 0; ; waitPeriod = int(math.Min(float64((waitPeriod+1)*2), float64(maxWaitPeriod))) {
-			waitDuration := time.Second * time.Duration(waitPeriod)
+		for waitPeriod := 0; ; waitPeriod = (waitPeriod + 1) * 2 {
```
Comment (Ricardo-Osorio) [inline watchers/common.go:27]: If you are not using the 3 params of a for loop statement you don't need it, instead
```go
waitPeriod := 0
for {
    ...
    waitPeriod = (waitPeriod + 1) * 2
}
```
Makes it a lot easier to read

File: controllers/storageoscluster/node-manager_operand.go [hunk 2]
```diff
@@ -187,16 +182,28 @@ func (c *NodeManagerOperand) PostReady(ctx context.Context, obj client.Object) e
...
 			}
 		}
 	})
 
+	go func() {
+		select {
+		case <-closeWatchWaiter:
+			return
+		case err := <-watchErrChan:
+			log.Error(err, "node watcher has some permanent issues")
```
Comment (Ricardo-Osorio) [inline controllers/storageoscluster/node-manager_operand.go:200]: "has some permanent issues" doesn't add any value to the msg
Also, the only err I see being passed to the chan is when `waitPeriod >= maxWaitPeriod`. If that's the case, add a meaningful log msg at that point and don't log anything here

File: main.go [hunk 3]
```diff
@@ -249,6 +249,11 @@ func main() {
 
 		return nil
 	})
+	go func() {
+		err := <-watchErrChan
+		setupLog.Error(err, "config map watcher has some permanent issues")
```
Comment (Ricardo-Osorio) [inline main.go:254]: Same as above

File: controllers/storageoscluster/node-manager_operand.go (hunk 2 above)
Comment (mhmxs) [reply controllers/storageoscluster/node-manager_operand.go:200]: not agree, at this point we don't know what is the reason. More details found in err. We can't relay on the implementation what we call. At this point we know, something went wrong. error will contain the reason.

File: main.go (hunk 3 above)
Comment (mhmxs) [reply main.go:254]: Same as above

File: watchers/common.go (hunk 1 above)
Comment (mhmxs) [reply watchers/common.go:27]: Not agree, this doesn't looks more readable:
```
waitPeriod := 0
for {
			if waitPeriod != 0 {
				log.Info("wait before start", "period", waitPeriod)
				time.Sleep(time.Second * time.Duration(waitPeriod))
			}

			err := watchFunc(ctx)
			if err != nil {
				if waitPeriod >= maxWaitPeriod {
					log.Error(err, "time to give up watching")
					errChan <- err
					break
				}

				log.Error(err, "there was an error during watch")
				waitPeriod = (waitPeriod + 1) * 2
				continue
			}

			// Watch should be stop.
			log.Info("watch has ended")
			close(errChan)
			break
		}
```

Comment (Ricardo-Osorio) [reply watchers/common.go:27]:
```go
waitPeriod := 0
for {
    if err := watchFunc(ctx); err == nil {
        log.Info("watch has ended")
        close(errChan) break
    }

    if watcherRecoveryPeriod >= maxWatcherRecoveryPeriod {
        log.Error(err, "watcher not able to recovered from failure")
        errChan <- err
        break
    }

    log.Error(err, "there was an error while watching...something") // add what this is watching
    waitPeriod = (waitPeriod + 1) * 2
    time.Sleep(time.Second * time.Duration(waitPeriod))
}
```

Comment (Ricardo-Osorio) [reply watchers/common.go:27]: This is what I meant, I also simplified the error handling to return earlier when possible and reducing the nesting code

File: watchers/common.go [hunk 4]
```diff
@@ -16,32 +15,41 @@ const maxWaitPeriod = 30
 type WatchConsumer func(<-chan watch.Event) error
 
 // watchFunc manages object watching.
-type watchFunc func(context.Context, time.Duration) error
+type watchFunc func(context.Context) error
 
 // start wathing given resources.
```
Comment (Ricardo-Osorio) [inline watchers/common.go:20]: If you don't mind fixing this type "wathing" -> "watching"

File: controllers/storageoscluster/node-manager_operand.go (hunk 2 above)
Comment (Ricardo-Osorio) [reply controllers/storageoscluster/node-manager_operand.go:200]: I had to look at the code locally to understand it better. I agree to not remove the logging but I stand with how I feel about "has some permanent issues" needing to be written in another way or removed

Comment (Ricardo-Osorio) [reply controllers/storageoscluster/node-manager_operand.go:200]: @nolancon @Mojachieee could you give your thoughts on this?
I am too picky with how log msgs should be

Comment (Mojachieee) [reply controllers/storageoscluster/node-manager_operand.go:200]: `config map watcher failed` seem good to me

Comment (nolancon) [reply controllers/storageoscluster/node-manager_operand.go]: isn't it a Daemonset watcher? 
Either way, "..... encountered error" gets my vote

**Review by Ricardo-Osorio — APPROVED** [review]

**Pull request merged**